File: src/programmers/skillcheck/level3/Coverage.java
package programmers.skillcheck.level3;

/**
 * 프로그래머스
 * 스킬체크 레벨 3
 *
 * Coverage
 * chanllenge_id=486
 *
 * 아파트 전체에 전파를 전달하기
 */
public class Coverage {

    public int solution(int n, int[] stations, int w) {
        // 하나의 기지국은 최대 2w + 1 만큼을 커버한다
        int maxCoverage = 2 * w + 1;

        int start = 1;
        int total = 0;

        for (int i = 0; i < stations.length; i++) {
            // start ~ stations[i] 까지의 기지국 수를 구한다
            int interval = stations[i] - w - start;
            total += (int) Math.ceil((double) interval / maxCoverage);

            start = stations[i] + w;
        }

        // 마지막 계산
        int interval = n - start;
        total += (int) Math.ceil((double) interval / maxCoverage);

        return total;
    }

    public static void main(String[] args) {
        Coverage obj = new Coverage();

        int n_ex1 = 11;
        int[] stations_ex1 = {4, 11};
        int w_ex1 = 1;
        obj.solution(n_ex1, stations_ex1, w_ex1);

        int n_ex2 = 16;
        int[] stations_ex2 = {9};
        int w_ex2 = 2;
        obj.solution(n_ex2, stations_ex2, w_ex2);
    }
}

/*
채점을 시작합니다.
정확성  테스트
테스트 1 〉	통과 (1.06ms, 52.7MB)
테스트 2 〉	통과 (1.05ms, 50.2MB)
테스트 3 〉	통과 (0.98ms, 53MB)
테스트 4 〉	실패 (1.88ms, 51.8MB)
테스트 5 〉	통과 (1.02ms, 50.3MB)
테스트 6 〉	통과 (1.04ms, 50.6MB)
테스트 7 〉	통과 (1.11ms, 50.5MB)
테스트 8 〉	실패 (1.61ms, 50.2MB)
테스트 9 〉	통과 (1.06ms, 52.4MB)
테스트 10 〉	통과 (1.08ms, 50.4MB)
테스트 11 〉	통과 (1.03ms, 50.8MB)
테스트 12 〉	통과 (0.91ms, 50.3MB)
테스트 13 〉	통과 (1.02ms, 52.4MB)
테스트 14 〉	통과 (1.11ms, 52MB)
테스트 15 〉	통과 (1.05ms, 50.6MB)
테스트 16 〉	통과 (1.06ms, 50MB)
테스트 17 〉	실패 (1.95ms, 50.4MB)
테스트 18 〉	통과 (1.10ms, 52.1MB)
테스트 19 〉	실패 (1.98ms, 52.2MB)
테스트 20 〉	통과 (1.09ms, 52.4MB)
테스트 21 〉	실패 (2.09ms, 52.3MB)
효율성  테스트
테스트 1 〉	통과 (4.98ms, 53.8MB)
테스트 2 〉	실패 (시간 초과)
테스트 3 〉	통과 (2.96ms, 55.2MB)
테스트 4 〉	실패 (시간 초과)
채점 결과
정확성: 26.8
효율성: 7.4
합계: 34.2 / 50
 */

File: src/programmers/skillcheck/level3/Coverage.js
/**
 * 프로그래머스
 * 스킬체크 레벨 3
 *
 * Coverage
 * chanllenge_id=486
 *
 * 아파트 전체에 전파를 전달하기
 */
function solution(n, stations, w) {

    // 한 개의 기지국이 커버할 수 있는 범위
    const coverage = 2 * w + 1;

    let start = 1;
    let end = n;
    let answer = 0;

    stations.forEach(st => {
        // start ~ 각 station 까지 범위의 최소 기지국 수를 계산
        // stations에서 커버된 범위를 뺌 st - w
        const interval = (st - w) - start;
        const cov = interval / coverage;

        // js에서 number/number가 어떻게 계산되는가..
        // 소수점 까지 계산되는군..
        // console.log(">>> cov : " + cov);
        // console.log(">>> Math.ceil(cov) : " + Math.ceil(cov));
        answer += Math.ceil(cov);

        start = st + w;
    });

    // 마지막 끝에서 마지막 기지국 계산
    const interval = end - (stations[stations.length - 1] + w);
    const cov = interval / coverage;
    answer += Math.ceil(cov);

    return answer;
}

/*
정확성  테스트
테스트 1 〉	통과 (1.69ms, 37.5MB)
테스트 2 〉	통과 (1.71ms, 37.4MB)
테스트 3 〉	통과 (1.72ms, 37.7MB)
테스트 4 〉	실패 (3.14ms, 37.9MB)
테스트 5 〉	통과 (1.69ms, 37.5MB)
테스트 6 〉	통과 (1.72ms, 37.5MB)
테스트 7 〉	통과 (1.71ms, 37.4MB)
테스트 8 〉	실패 (3.14ms, 37.8MB)
테스트 9 〉	통과 (1.74ms, 37.4MB)
테스트 10 〉	통과 (1.74ms, 37.4MB)
테스트 11 〉	통과 (1.73ms, 37.5MB)
테스트 12 〉	통과 (1.70ms, 37.4MB)
테스트 13 〉	통과 (1.74ms, 37.6MB)
테스트 14 〉	통과 (1.70ms, 37.4MB)
테스트 15 〉	통과 (1.70ms, 37.5MB)
테스트 16 〉	통과 (1.72ms, 37.5MB)
테스트 17 〉	실패 (3.10ms, 37.8MB)
테스트 18 〉	통과 (1.72ms, 37.4MB)
테스트 19 〉	실패 (3.16ms, 37.7MB)
테스트 20 〉	통과 (1.70ms, 37.5MB)
테스트 21 〉	실패 (3.13ms, 37.8MB)
효율성  테스트
테스트 1 〉	통과 (3.28ms, 38.3MB)
테스트 2 〉	통과 (3.22ms, 38.2MB)
테스트 3 〉	통과 (3.28ms, 38.4MB)
테스트 4 〉	통과 (3.22ms, 38.2MB)
채점 결과
정확성: 26.8
효율성: 14.8
합계: 41.6 / 50
 */